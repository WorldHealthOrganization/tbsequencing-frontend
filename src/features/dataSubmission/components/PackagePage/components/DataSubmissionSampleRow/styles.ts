import { css } from '@emotion/react';
import appColors from '../../../../../../styles/colors';

const VERTICAL_PADDING = 9;

const VERTICAL_PADDING_PROPS = {
  paddingTop: VERTICAL_PADDING,
  paddingBottom: VERTICAL_PADDING,
  paddingLeft: 19,
  paddingRight: 30,
};

export const container = css({
  ...VERTICAL_PADDING_PROPS,
  boxShadow: 'inset 0px -1px 0px rgba(0, 0, 0, 0.12)',
});

export const info = css({
  display: 'flex',
  alignItems: 'center',
});

export const actionsContainer = css({
  marginLeft: 'auto',
});

export const infoContainer = css({
  marginLeft: 19,
});

export const error = css({
  backgroundColor: appColors.status.emergencyRedBg,
});

export const sxIconDone = {
  fill: appColors.secondary.whoBlue,
  fontSize: 17,
};

export const iconDisabledDowload = {
  fill: appColors.neutral.ncbiGrey,
  fontSize: 17,
};

export const sxIconDoneWarning = {
  fill: appColors.secondary.whoOrange,
  fontSize: 17,
};

export const sxIconDangerous = {
  fill: appColors.status.whoEmergencyRed,
  fontSize: 22,
};

export const errorText = css({
  color: appColors.status.whoEmergencyRed,
});

export const regularText = css({
  color: appColors.tints.whoBlue.light,
});

export const warningText = css({
  color: appColors.secondary.whoOrange,
});
