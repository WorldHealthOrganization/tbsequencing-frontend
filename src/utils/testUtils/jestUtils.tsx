import React, { ReactElement } from 'react';
import { render as rtlRender } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';

import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import { reducer } from '../../app/store';

const getWrapper = (
  ui: ReactElement,
  {
    // @ts-ignore
    preloadedState,
    store = configureStore({ reducer, preloadedState }),
  } = {},
) => ({ children }: { children: ReactElement }) => (
  <Provider store={store}>{children}</Provider>
);

function render(
  ui: ReactElement,
  {
    // @ts-ignore
    preloadedState,
    store = configureStore({ reducer, preloadedState }),
    ...renderOptions
  } = {},
) {
  const Wrapper = getWrapper(ui, { preloadedState, store });

  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

const getTreeJSONWithRedux = (ui: ReactElement, {
  // @ts-ignore
  preloadedState,
  store = configureStore({ reducer, preloadedState }),
} = {}) => {
  const Wrapper = getWrapper(ui, { preloadedState, store });

  const tree = renderer
    .create(<Wrapper>{ui}</Wrapper>)
    .toJSON();

  return tree;
};

export const wrappedInRouter = (
  component: ReactElement,
) => <BrowserRouter>{component}</BrowserRouter>;

export * from '@testing-library/react';
export { render, getTreeJSONWithRedux };
