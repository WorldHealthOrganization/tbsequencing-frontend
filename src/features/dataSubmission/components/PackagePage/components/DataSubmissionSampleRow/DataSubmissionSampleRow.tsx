/** @jsxImportSource @emotion/react */
import React, { ReactElement, useState } from 'react';
import {
  CircularProgress, IconButton, Popover, Typography,
} from '@mui/material';
import {
  Done, DangerousOutlined, Close, FileDownload,
} from '@mui/icons-material';
import { match } from 'ts-pattern';
import { SerializedStyles } from '@emotion/react';
import * as styles from './styles';
import PrimaryText from '../../../../../../components/typography/PrimaryText';

interface Props {
  isError?: boolean;
  isWarning?: boolean;
  title: string;
  description?: string;
  isLoading?: boolean;
  onRemoveClick?: () => void;
  onDownloadClick?: () => void;
  containerStyle?: SerializedStyles;
  controlDisabled: boolean;
  isSomethingInSequencingDataLoading?: boolean;
  isUloadingFile?: boolean;
  isProcessingFile?: boolean;
  dataLocation?: string;
}

interface IMatchIcon {
  error: boolean;
  loading: boolean;
  warning: boolean;
  isUloadingFile?: boolean;
  isProcessingFile?: boolean;
}

type MatchDescriptionStyle = Omit<IMatchIcon, 'loading'>;

export const DataSubmissionSampleRow = ({
  isError, isLoading, onDownloadClick, onRemoveClick,
  description, title, containerStyle, controlDisabled,
  isWarning, isSomethingInSequencingDataLoading, isUloadingFile, isProcessingFile, dataLocation,
}: Props) => {
  const errorStyles = isError && styles.error;
  const error = Boolean(isError);
  const loading = Boolean(isLoading);
  const warning = Boolean(isWarning);
  const [anchorEl, setAnchorEl] = useState<HTMLButtonElement | null>(null);

  const icon = match<IMatchIcon, ReactElement>({
    error, loading, warning, isUloadingFile, isProcessingFile,
  })
    .with({ isUloadingFile: false, isProcessingFile: true }, () => <CircularProgress size={14} />)
    .with({
      error: false, loading: false, warning: false, isUloadingFile: false, isProcessingFile: false,
    }, () => <Done sx={styles.sxIconDone} />)
    .with({ error: true, loading: false }, () => <DangerousOutlined sx={styles.sxIconDangerous} />)
    .with({ error: false, isUloadingFile: true }, () => <CircularProgress size={14} />)
    .with({
      error: false,
      loading: false,
      warning: true,
    }, () => <Done sx={styles.sxIconDoneWarning} />)
    .otherwise(() => <Done sx={styles.sxIconDone} />);

  const descriptionStyle = match<MatchDescriptionStyle, SerializedStyles>({ error, warning })
    .with({ warning: true, error: false }, () => styles.warningText)
    .with({ warning: false, error: true }, () => styles.errorText)
    .otherwise(() => styles.regularText);

  const actionsDisabled = isLoading || controlDisabled || isSomethingInSequencingDataLoading;

  const shouldRenderDownloadButton = Boolean(onDownloadClick) && dataLocation !== 'NCBI';
  const shouldRenderDisabledDownloadButton = Boolean(onDownloadClick) && dataLocation === 'NCBI';

  const shouldRenderRemoveButton = Boolean(onRemoveClick)
  && !controlDisabled
  && !isUloadingFile
  && !isProcessingFile;

  const handleClickHint = (event: React.MouseEvent<HTMLButtonElement>) => {
    setAnchorEl(event.currentTarget);
  };

  const handleCloseHint = () => {
    setAnchorEl(null);
  };

  const open = Boolean(anchorEl);
  const id = open ? 'upload-popover' : undefined;

  return (
    <div css={[styles.container, errorStyles, containerStyle]}>
      <div css={styles.info}>
        {icon}
        <div css={styles.infoContainer}>
          <div><PrimaryText>{title}</PrimaryText></div>
          {description
          && <div><PrimaryText style={descriptionStyle}>{description}</PrimaryText></div>}
        </div>
        <div css={styles.actionsContainer}>
          {shouldRenderDownloadButton && (
          <IconButton
            data-testid="download-btn"
            onClick={onDownloadClick}
          >
            <FileDownload sx={styles.sxIconDone} />
          </IconButton>
          )}
          {shouldRenderDisabledDownloadButton && (
          <IconButton
            data-testid="disabled-download-btn"
            aria-label="toggle download info"
            aria-describedby={id}
            onClick={handleClickHint}
          >
            <FileDownload sx={styles.iconDisabledDowload} />
          </IconButton>
          )}
          <Popover
            anchorOrigin={{
              vertical: 'center',
              horizontal: 'left',
            }}
            transformOrigin={{
              vertical: 'center',
              horizontal: 'right',
            }}
            id={id}
            open={open}
            anchorEl={anchorEl}
            onClose={handleCloseHint}
          >
            <Typography sx={{ p: 2 }}>
              The file has been discarded in favor of NCBI SRA record found
            </Typography>
          </Popover>
          {shouldRenderRemoveButton && (
          <IconButton
            disabled={actionsDisabled}
            data-testid="remove-btn"
            onClick={onRemoveClick}
          >
            <Close sx={styles.sxIconDone} />
          </IconButton>
          )}
        </div>
      </div>
    </div>
  );
};

export default DataSubmissionSampleRow;
