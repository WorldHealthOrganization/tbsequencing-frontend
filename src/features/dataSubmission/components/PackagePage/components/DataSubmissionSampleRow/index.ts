import DataSubmissionSampleRow from './DataSubmissionSampleRow';

export default DataSubmissionSampleRow;
