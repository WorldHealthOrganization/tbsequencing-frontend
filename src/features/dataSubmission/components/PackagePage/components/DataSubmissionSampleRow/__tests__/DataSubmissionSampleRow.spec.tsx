import React from 'react';
import { screen } from '@testing-library/react';
import { getTreeJSONWithRedux, render } from '../../../../../../../utils/testUtils/jestUtils';
import DataSubmissionSampleRow from '../DataSubmissionSampleRow';

describe('DataSubmissionSampleRow component test', () => {
  const title = 'mockTitle';
  const mockDescription = 'mockDescription';
  const mockOnRemoveClick = jest.fn();
  const mockControlsDisabled = false;

  test('renders DataSubmissionSampleRow component without error', () => {
    render(
      <DataSubmissionSampleRow
        title={title}
        isLoading={false}
        description={mockDescription}
        onRemoveClick={mockOnRemoveClick}
        controlDisabled={mockControlsDisabled}
      />,
    );
  });

  it('Should render title properly', () => {
    render(
      <DataSubmissionSampleRow
        title={title}
        isLoading={false}
        description={mockDescription}
        onRemoveClick={mockOnRemoveClick}
        controlDisabled={mockControlsDisabled}
      />,
    );

    screen.getByText(title);
  });

  it('Should render description properly', () => {
    render(
      <DataSubmissionSampleRow
        title={title}
        isLoading={false}
        description={mockDescription}
        onRemoveClick={mockOnRemoveClick}
        controlDisabled={mockControlsDisabled}
      />,
    );

    screen.getByText(mockDescription);
  });

  it('Should not render downloadButton without callback prop', () => {
    render(
      <DataSubmissionSampleRow
        title={title}
        isLoading={false}
        description={mockDescription}
        onRemoveClick={mockOnRemoveClick}
        controlDisabled={mockControlsDisabled}
      />,
    );

    const foundButton = screen.queryByTestId('download-btn');

    expect(foundButton).toBeNull();
  });

  it('Should call onDownload callback on buttonClick', () => {
    const mockOnDownloadHandler = jest.fn();

    render(
      <DataSubmissionSampleRow
        title={title}
        isLoading={false}
        description={mockDescription}
        onRemoveClick={mockOnRemoveClick}
        onDownloadClick={mockOnDownloadHandler}
        controlDisabled={mockControlsDisabled}
      />,
    );

    const downloadBtn = screen.getByTestId('download-btn');
    downloadBtn.click();

    expect(mockOnDownloadHandler).toBeCalled();
  });

  it('Should call on remove callback on remove button click', () => {
    render(
      <DataSubmissionSampleRow
        title={title}
        isLoading={false}
        description={mockDescription}
        onRemoveClick={mockOnRemoveClick}
        controlDisabled={mockControlsDisabled}
      />,
    );

    const removeBtn = screen.getByTestId('remove-btn');
    removeBtn.click();

    expect(mockOnRemoveClick).toBeCalled();
  });

  it('Should render correctly', () => {
    const tree = getTreeJSONWithRedux(
      <DataSubmissionSampleRow
        title={title}
        isLoading={false}
        description={mockDescription}
        onRemoveClick={mockOnRemoveClick}
        controlDisabled={mockControlsDisabled}
      />,
    );

    expect(tree).toMatchSnapshot();
  });
});
