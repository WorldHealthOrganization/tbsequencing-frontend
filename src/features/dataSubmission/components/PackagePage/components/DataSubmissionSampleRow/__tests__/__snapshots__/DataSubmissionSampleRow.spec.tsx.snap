// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataSubmissionSampleRow component test Should render correctly 1`] = `
<div
  className="css-10zynoi-DataSubmissionSampleRow"
>
  <div
    className="css-yb5sry-DataSubmissionSampleRow"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-2rraxu-MuiSvgIcon-root"
      data-testid="DoneIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M9 16.2 4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
      />
    </svg>
    <div
      className="css-1cp60dy-DataSubmissionSampleRow"
    >
      <div>
        <p
          className="MuiTypography-root MuiTypography-body1 css-vtlwgn-MuiTypography-root-BaseText"
        >
          mockTitle
        </p>
      </div>
      <div>
        <p
          className="MuiTypography-root MuiTypography-body1 css-1y45chy-MuiTypography-root-BaseText"
        >
          mockDescription
        </p>
      </div>
    </div>
    <div
      className="css-1xzex0e-DataSubmissionSampleRow"
    >
      <button
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        data-testid="remove-btn"
        disabled={false}
        onBlur={[Function]}
        onClick={[MockFunction]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-2rraxu-MuiSvgIcon-root"
          data-testid="CloseIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
